odoo/api.py

```diff
@@ -854,6 +854,28 @@ def user(self):
         """ return the current user (as an instance) """
         return self(user=SUPERUSER_ID)['res.users'].browse(self.uid)
 
+    @property
+    def company_id(self):
+        """ return the company in which the user is logged in (as an instance) """
+        if self.context.get('allowed_company_ids'):
+            company_id = int(self.context.get('allowed_company_ids').split(',')[0])
+            if company_id in self.user.company_ids.ids:
+                return self['res.company'].browse(company_id)
+        return self.user.company_id
+
+    @property
+    def company_ids(self):
+        """ return a recordset of the enabled companies by the user """
+        try:  # In case the user tries to bidouille the url (eg: cids=1,foo,bar)
+            users_company_ids = self.user.company_ids.ids
+            cids = self.context.get('allowed_company_ids')
+            allowed_company_ids = [int(company_id) for company_id in cids.split(',') if company_id]
+            # Prevent the user to enable companies for which he doesn't have any access
+            allowed_company_ids = [company_id for company_id in allowed_company_ids if company_id in users_company_ids]
+        except:
+            allowed_company_ids = [self.company_id.id]
```

Comment:
If we change the default to all user companies (user.allowed_company_ids), we save a lot of potential trouble in all "out of context" calls, such as /mail/redirect or /web/image, etc. And it is not unsafe because the user *does* have access to these other companies. The risk of exposing foreign records (wrt to the context) is low because all normal RPCs *will have* a proper allowed_company_ids.
- change the default here
- add a comment to explain why we do it (avoid needless access problems)
- remove the patches where they are not needed anymore (typically when we used the str join technique inside controllers like /mail/redirect or /web/image)